xla/service/gpu/cudnn_fused_mha_rewriter.cc

```diff
@@ -15,44 +15,90 @@ limitations under the License.
 
 #include "xla/service/gpu/cudnn_fused_mha_rewriter.h"
 
-#include <numeric>
-#include <optional>
+#include <functional>
+#include <queue>
 #include <string>
-#include <utility>
-#include <vector>
 
+#include "tsl/platform/errors.h"
+#include "tsl/platform/statusor.h"
 #include "xla/hlo/ir/hlo_casting_utils.h"
 #include "xla/hlo/ir/hlo_instruction.h"
+#include "xla/literal_util.h"
 #include "xla/permutation_util.h"
+#include "xla/primitive_util.h"
 #include "xla/service/gpu/backend_configs.pb.h"
 #include "xla/service/gpu/cublas_cudnn.h"
+#include "xla/service/gpu/ir_emission_utils.h"
 #include "xla/service/gpu/matmul_utils.h"
+#include "xla/service/hlo_creation_utils.h"
 #include "xla/service/pattern_matcher.h"
+#include "xla/stream_executor/dnn.pb.h"
 #include "xla/stream_executor/stream_executor.h"
-#include "tsl/platform/errors.h"
-#include "tsl/platform/statusor.h"
 
 namespace xla {
 namespace gpu {
 namespace {
 namespace m = match;
 
+// A struct that contains all the matched nodes
+// and results from pattern matching forward graph
+struct MatchFwdResult {
+  HloInstruction* matched_bmm_1 = nullptr;
+  HloInstruction* matched_bmm_2 = nullptr;
+  HloInstruction* matched_bias = nullptr;
+  HloInstruction* matched_mask = nullptr;
+  HloInstruction* matched_scale = nullptr;
+  HloInstruction* matched_softmax_input = nullptr;
+
+  double matched_dropout_rate = 0.0;
+  bool need_canonicalization = false;
+  bool is_training = false;
+  bool has_match = false;
+  std::string matched_custom_call_name;
+};
+
+// A struct that contains all the matched nodes
+// and results from pattern matching backward graph
+struct MatchBwdResult {
+  HloInstruction* matched_bmm_1_grad_1 = nullptr;
+  HloInstruction* matched_bmm_1_grad_2 = nullptr;
+
+  HloInstruction* matched_bmm_2_grad_1 = nullptr;
+  HloInstruction* matched_bmm_2_grad_2 = nullptr;
+  HloInstruction* matched_d_intermediate = nullptr;
+  // We use this to keep track of all gradient bmms that need
+  // canonicalization.
+  bool bmm_1_grad_1_need_canonicalization = false;
+  bool bmm_1_grad_2_need_canonicalization = false;
+  bool bmm_2_grad_1_need_canonicalization = false;
+  bool bmm_2_grad_2_need_canonicalization = false;
+
+  bool has_match = false;
+  std::string matched_custom_call_name;
+};
+
 template <typename Pattern>
 auto OptionalReshape(Pattern pattern) {
   auto shared = m::SharedSubpattern(pattern);
-  return m::AnyOf<HloInstruction>(m::Reshape(shared), shared);
+  return m::AnyOf<HloInstruction>(m::Reshape(pattern), shared);
```

Comment:
This should be: 
```
auto shared = m::SharedSubpattern(pattern);
return m::AnyOf<HloInstruction>(m::Reshape(shared), shared);
```

If you write (m::Reshape(pattern), shared); then the compilation can be slow, because the shared pattern is not reused.

Was there a problem with the original code?

Reply:
ah this was mostly an error from rebase, let me change it back.